<a href="https://colab.research.google.com/github/embob1/ML-fundamentals-2026/blob/main/assignment_1_MOHAMED_EMBABY.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()


Saving bank+marketing.zip to bank+marketing.zip


In [3]:
import zipfile

with zipfile.ZipFile("bank+marketing.zip", "r") as zip_ref:
    zip_ref.extractall()

print("Unzipped successfully!")

Unzipped successfully!


In [4]:
import os
os.listdir()

['.config',
 'bank.zip',
 'bank-additional.zip',
 'bank+marketing.zip',
 'sample_data']

In [6]:
with zipfile.ZipFile("bank-additional.zip", "r") as zip_ref:
    zip_ref.extractall()

print("Second unzip done!")

Second unzip done!


In [7]:
os.listdir()


['.config',
 'bank.zip',
 'bank-additional.zip',
 'bank-additional',
 '__MACOSX',
 'bank+marketing.zip',
 'sample_data']

In [8]:
os.listdir("bank-additional")

['bank-additional-full.csv',
 '.DS_Store',
 'bank-additional.csv',
 'bank-additional-names.txt',
 '.Rhistory']

In [9]:
df = pd.read_csv("bank-additional/bank-additional.csv", sep=";")

print("Shape:", df.shape)
df.head()

Shape: (4119, 21)


,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,30,blue-collar,married,basic.9y,no,yes,no,cellular,may,fri,...,2,999,0,nonexistent,-1.8,92.893,-46.2,1.313,5099.1,no
1,39,services,single,high.school,no,no,no,telephone,may,fri,...,4,999,0,nonexistent,1.1,93.994,-36.4,4.855,5191.0,no
2,25,services,married,high.school,no,yes,no,telephone,jun,wed,...,1,999,0,nonexistent,1.4,94.465,-41.8,4.962,5228.1,no
3,38,services,married,basic.9y,no,unknown,unknown,telephone,jun,fri,...,3,999,0,nonexistent,1.4,94.465,-41.8,4.959,5228.1,no
4,47,admin.,married,university.degree,no,yes,no,cellular,nov,mon,...,1,999,0,nonexistent,-0.1,93.200,-42.0,4.191,5195.8,no


In [10]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

The dataset contains 4,119 observations and 21 variables.

The dataset includes both categorical and numerical features.

Categorical variables describe demographic and campaign-related information (e.g., job, marital status, education, contact type).

Numerical variables include age, campaign information, and macroeconomic indicators.

The target variable y indicates whether the client subscribed to a term deposit.bold text **bold text**

In [11]:
df['y'].value_counts()

,count
y,
no,3668
yes,451


In [12]:
df['y'].value_counts(normalize=True)

,proportion
y,
no,0.890507
yes,0.109493


The target variable y is imbalanced.

Approximately 89% of clients did not subscribe to a term deposit, while only 11% subscribed.

This indicates a significant class imbalance, which may affect model training and evaluation.

In particular, accuracy alone would not be an appropriate metric, since a model predicting only “no” would already achieve around 89% accuracy. **bold text**

In [13]:
from sklearn.model_selection import train_test_split

# Separate features and target
X = df.drop("y", axis=1)
y = df["y"]

# First split: Train (60%) and Temp (40%)
X_train, X_temp, y_train, y_temp = train_test_split(
    X, y,
    test_size=0.4,
    stratify=y,
    random_state=42
)

# Second split: Validation (20%) and Test (20%)
X_val, X_test, y_val, y_test = train_test_split(
    X_temp, y_temp,
    test_size=0.5,
    stratify=y_temp,
    random_state=42
)

print("Train shape:", X_train.shape)
print("Validation shape:", X_val.shape)
print("Test shape:", X_test.shape)

Train shape: (2471, 20)
Validation shape: (824, 20)
Test shape: (824, 20)


Data Splitting Strategy

The dataset was split into:

60% training set

20% validation set

20% test set

Stratified splitting was used to preserve the original class distribution of the target variable.

Since the dataset is imbalanced (approximately 89% “no” and 11% “yes”), stratification ensures that all subsets maintain similar proportions.

Without stratification, the class distribution could differ across splits, leading to unreliable model evaluation.

In [14]:
X_train.isnull().sum()

,0
age,0
job,0
marital,0
education,0
default,0
housing,0
loan,0
contact,0
month,0
day_of_week,0


In [15]:
(X_train == "unknown").sum()

,0
age,0
job,20
marital,7
education,104
default,492
housing,55
loan,55
contact,0
month,0
day_of_week,0


Missing Values Analysis

The dataset does not contain explicit missing values (no NaN values were detected).

However, several categorical variables contain the category "unknown", which represents implicit missing values.

The variables affected include:

job

marital

education

default

housing

loan

In this assignment, "unknown" will be treated as a valid category rather than being removed or imputed.

Removing these rows would unnecessarily reduce the dataset size, and imputing them could introduce bias.

When categorical encoding is applied later, "unknown" will naturally be handled as an additional category.

In [1]:
File  Save a copy in GitHub

SyntaxError: invalid character '→' (U+2192) (3546989898.py, line 1)